### PR TITLE
Support for TUI

### PR DESCRIPTION
**Pull request by pwang00**
Added an additional thread in manticore/core/worker.py to set up socket server and communicate with TUI via serialized protobuf messages.  Currently more of a proof-of-concept than a robust implementation, but hopefully adding additional functionality won't be too difficult.

Current setup has hardcoded host and port (127.0.0.1, 1337), and successfully sends serialized states to the client TUI.  In the future, the host and port should be saved by ansible upon provisioning a DO droplet.

### CODE REVIEW

**Comment by ehennenfent** [comment]
Closes #1272

**Comment by ekilmer** [comment]
As a general rule, generated files shouldn't be checked into the repo. They should be generated when the project is built. This way, there is no chance of forgetting to update them manually.

**Comment by pwang00** [comment]
> As a general rule, generated files shouldn't be checked into the repo. They should be generated when the project is built. This way, there is no chance of forgetting to update them manually.

Yeah point taken.  Atm I think the Manticore dependency list has to be updated to account for protobuf and npyscreen.

**Comment by ekilmer** [comment]
Regarding the generation of protobuf files during build: Unfortunately, I don't know the best way of doing this for a python project, but I think taking a look at Google's own [`setup.py`](https://github.com/protocolbuffers/protobuf/blob/master/python/setup.py) might be a good place to start.

I can also help with taking a closer look if nothing helpful is turning up 👍

**Comment by pwang00** [comment]
Actually looking into this right now, seems like a possible way to do this is to just run a post-install script via setuptools.  I'll test and push if this ends up working.

File: setup.py [hunk 1]
```diff
@@ -34,6 +36,7 @@ def rtd_dependent_deps():
     "lint": lint_deps,
 }
 
+subprocess.Popen(protoc_cmd)
```
Comment (ekilmer) [inline setup.py]: We should use `subprocess.checkout_output` for this https://docs.python.org/3/library/subprocess.html#subprocess.check_output

File: setup.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def rtd_dependent_deps():
     "lint": lint_deps,
 }
 
-subprocess.Popen(protoc_cmd)
+output = subprocess.check_output(protoc_cmd)
```
Comment (ekilmer) [inline setup.py]: Nit: If we don't use the `output` variable, then let's just get rid of it. The `check_output` function will throw a `CalledProcessError` exception when the exit code is non-zero, which should be fine in this case

**Comment by ehennenfent** [comment]
> As a general rule, generated files shouldn't be checked into the repo. They should be generated when the project is built. This way, there is no chance of forgetting to update them manually.

Looks like `pip install protobuf` doesn't install the `protoc` compiler. That leaves us with two options:
* Add `protoc` as an external dependency that must be installed via `apt` before installing Manticore
* Add the generated `state_pb2.py` file back to the repository and rely on our CI to tell us when to update it

Personally I'd vote for the second option.

**Comment by pwang00** [comment]
> > As a general rule, generated files shouldn't be checked into the repo. They should be generated when the project is built. This way, there is no chance of forgetting to update them manually.
> 
> Looks like `pip install protobuf` doesn't install the `protoc` compiler. That leaves us with two options:
> 
>     * Add `protoc` as an external dependency that must be installed via `apt` before installing Manticore
> 
>     * Add the generated `state_pb2.py` file back to the repository and rely on our CI to tell us when to update it
> 
> 
> Personally I'd vote for the second option.

Yeah, the user would have to (on ubuntu) `sudo apt-get install protobuf-compiler` and I suppose depending on the distribution (e.g. Amazon AMI) even finding the right package name to install could be a pain. I don't think there's an efficient way to automatically install protoc for any given Linux distro.

I would probably go with the second option as well since I think doing so would save the user a step in the installation process, and that the protocol shouldn't change *that* often across builds.

File: manticore/core/manticore.py [hunk 3]
```diff
@@ -191,7 +191,7 @@ def __init__(self, initial_state, workspace_url=None, policy="random", **kwargs)
 
         During exploration Manticore spawns a number of temporary states that are
         maintained in different lists:
-
+
```
Comment (ekilmer) [inline manticore/core/manticore.py]: Looks like some dangling whitespace made it in here somehow?

**Pull request merged**